cli/cmd/test_connection.go

```diff
@@ -171,34 +186,146 @@ func testConnection(cmd *cobra.Command, args []string) error {
+
+	maxLength := 0
+	for _, testResult := range testConnectionResults {
+		if len(testResult.pluginRef) > maxLength {
+			maxLength = len(testResult.pluginRef)
+		}
+	}
+
+	connFailures := make([]testConnectionResult, 0, len(testConnectionResults))
+	for i, testResult := range testConnectionResults {
+		if i == 0 {
+			cmd.Println("Test Connection Results")
+		}
+		cmd.Printf("%-12s %s %s", testResult.pluginKind, testResult.pluginRef, strings.Repeat(" ", maxLength-len(testResult.pluginRef)+1))
+		if testResult.Success {
+			cmd.Println("Success")
+			continue
+		}
+		cmd.Println("Failure:", testResult.FailureCode, "\t", testResult.FailureDescription)
+		connFailures = append(connFailures, testConnectionResults[i])
+	}
+
+	if len(connFailures) > 0 {
+		allErrors = errors.Join(allErrors, &testConnectionFailures{failed: connFailures})
+	}
 
 	return allErrors
 }
+
+type testConnectionFailures struct {
+	failed []testConnectionResult
+}
+
+func (*testConnectionFailures) Error() string {
+	// The errs are already shown in the console, so we don't need to show them again here
+	return "at least one test connection failed"
+}
+
+type testConnectionResult struct {
+	pluginRef          string
+	pluginKind         string
+	Success            bool
+	FailureCode        string `json:",omitempty"`
+	FailureDescription string `json:",omitempty"`
+}
+
+func testPluginConnection(ctx context.Context, client plugin.PluginClient, spec map[string]any) (*testConnectionResult, error) {
+	specBytes, err := marshalSpec(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	in := &plugin.TestConnection_Request{
+		Spec: specBytes,
+	}
+
+	resp, err := client.TestConnection(ctx, in)
+	if err != nil {
+		if gRPCErr, ok := grpcstatus.FromError(err); ok {
+			if gRPCErr.Code() == codes.Unimplemented {
+				err := initPlugin(ctx, client, spec, false, invocationUUID.String())
+				if err != nil {
+					return &testConnectionResult{
+						Success:            false,
+						FailureCode:        string(sdkPlugin.TestConnFailureCodeUnknown),
+						FailureDescription: err.Error(),
+					}, nil
+				}
+
+				return &testConnectionResult{
+					Success: true,
+				}, nil
+			}
+		}
+		return nil, err
+	}
+
+	return &testConnectionResult{
+		Success:            resp.Success,
+		FailureCode:        resp.FailureCode,
+		FailureDescription: resp.FailureDescription,
+	}, nil
+}
+
+// filterFailedTestResults fetch the failed test results.
+//
+// The function returns any failed test results, or nil if all tests passed. The hackernews plugin is excluded from the
+// failed test results since it was previously used as a test case for the test connection command.
+func filterFailedTestResults(results []testConnectionResult) (*testConnectionResult, error) {
+	var failedResults []testConnectionResult
+
+	for _, result := range results {
+		if !result.Success {
+			if strings.Contains(result.pluginRef, "cloudquery/file v4.0.4") || strings.Contains(result.pluginRef, "cloudquery/hackernews v3.0.25") {
```

Comment:
It's in the format `cloudquery/gandi@v3.2.1` I think my comment was misleading.
```suggestion
			if strings.Contains(result.pluginRef, "cloudquery/file@v4.0.4") || strings.Contains(result.pluginRef, "cloudquery/hackernews@v3.0.25") {
```